client/src/includes/stimulus.test.js

```diff
@@ -0,0 +1,253 @@
+    this.updateCount();
+  },
+  setupOutput() {
+    if (this.output) return;
+    const template = document.createElement('template');
+    template.innerHTML = `<output name='word-count' for='${this.element.id}' style='float: right;'></output>`;
+    const output = template.content.firstChild;
+    this.element.insertAdjacentElement('beforebegin', output);
+    this.output = output;
+  },
+  updateCount(event) {
+    const value = event ? event.target.value : this.element.value;
+    const words = (value || '').split(' ');
+    this.output.textContent = `${words.length} / ${this.maxValue} words`;
+  },
+  disconnect() {
+    this.output && this.output.remove();
+  },
+};
+
+/**
+ * Example controller from documentation as an ES6 class
+ */
+class WordCountController extends Controller {
+  static values = { max: { default: 10, type: Number } };
+
+  connect() {
+    const output = document.createElement('output');
+    output.setAttribute('name', 'word-count');
+    output.setAttribute('for', this.element.id);
+    output.style.float = 'right';
+    this.element.insertAdjacentElement('beforebegin', output);
+    this.output = output;
+    this.updateCount();
+  }
+
+  setupOutput() {
+    if (this.output) return;
+    const template = document.createElement('template');
+    template.innerHTML = `<output name='word-count' for='${this.element.id}' style='float: right;'></output>`;
+    const output = template.content.firstChild;
+    this.element.insertAdjacentElement('beforebegin', output);
+    this.output = output;
+  }
+
+  updateCount(event) {
+    const value = event ? event.target.value : this.element.value;
+    const words = (value || '').split(' ');
+    this.output.textContent = `${words.length} / ${this.maxValue} words`;
+  }
+
+  disconnect() {
+    this.output && this.output.remove();
+  }
+}
+
+describe('initStimulus', () => {
+  document.body.innerHTML = `
+  <main>
+  <section data-controller="w-test-mock">
+    <div id="item" data-w-test-mock-target="item"></div>
+  </section>
+  </main>`;
+
+  const mockControllerConnected = jest.fn();
+
+  class TestMockController extends Controller {
+    static targets = ['item'];
+
+    connect() {
+      mockControllerConnected();
+      this.itemTargets.forEach((item) => {
+        item.setAttribute('hidden', '');
+      });
+    }
+  }
+
+  const definitions = [
+    { identifier: 'w-test-mock', controllerConstructor: TestMockController },
+  ];
+
+  let application;
+
+  // note: no Wagtail code should load before this, however it is good to ensure this event fires
+  const stimulusInit = jest.fn();
+  document.addEventListener('wagtail:stimulus-init', stimulusInit);
+
+  const stimulusReady = jest.fn(({ detail }) => {
+    // register example controllers for other tests
+    detail.register('example-a', detail.createController(wordCountController));
+    detail.register('example-b', WordCountController);
+  });
+
+  document.addEventListener('wagtail:stimulus-ready', stimulusReady);
```

Comment:
Here and above – I think this code should be in a `beforeEach` or `beforeAll`? Jest can run tests in parallel, I’m not sure what happens when using globals like `document.body` at the top level like this.

Reply:
Fixed